### PR TITLE
[MRG] Update GUI to latest version of ipywidgets and voila

### PR DESCRIPTION
**Pull request by gtdang**
**Completed**
* Removed pinning of older versions ipywidgets, ipympl, and voila. Set a requirement for ipywidgets >=8.1.1, because of breaking changes to v8.x. 
    * Updating to the latest versions resolved js errors that were previously preventing the GUI from rendering, which was the impetus for pinning versions #585 #586. 
* Addressed some [ipywidgets 8.x API changes](https://ipywidgets.readthedocs.io/en/latest/user_migration_guides.html) that prevented GUI composition, running simulations, and plotting. 
    * Implemented new methods of assigning titles and children to collection widgets (Tabs and Accordion). ~~Though only to the scope of what was causes crashes with what I was testing (composition and simulation launch).~~ There could be more areas were it should be updated. 
    * Linking of Tab collections was causinging an issue with new ipywidgets validation logic. Solution was to write a decorator function that unlinks linked Tab collections (Figure tabs and Fig-Config tab groups) before editing functions and re-links them upon completion. 
        * Added test for decorator.
    * Updated to new `FileUpload` widget API
        * This involved restructuring the input structure to the widget. It now takes a list of dicts instead of a single dict.
        * Access to the file upload dict items was updated.
        * For the file upload pytests the simulated file dict also had to be restructured to the new dictionary structure FileUpload expects.
        * Removed setting of the deprecated `_counter` trait of `FileUpload`.
    * Added `ipykernel` as a GUI dependency. ipywidgets removed its dependency of ipykernel to allow more flexibility to use other kernels. This might only need to be a test dependency? 

**Remarks**
* ~~I've only tested the using the GUI to launch the default simulation and adding some of the plots. Will need to test/search for more updates to the ipywidgets 8.x API. I suspect features like file-upload also need to be updated.~~

**Questions**
* The `FileUpload` button widget has a counter displayed showing the number of files uploaded. This counter cannot be disabled. (https://github.com/jupyter-widgets/ipywidgets/issues/2904) For the network connectivity and drives upload button, the current code always keeps this number at 0. Do we want to keep it that way or should it counter increment to 1 if the user loads a param file? 
    * ~~Personally I don't think the counter is great for our application and wish it could be removed... I think the user does need to have some confirmation that a file has been loaded though. An informational display of what files have been loaded for network and drive would be nice. And this could even show that a default file has been loaded at the start.~~ We decided to keep the counter always at 0. 
    * On a slightly related note, this widget is able to accept multiple files. Would this be useful for loading multiple data files? That could be a feature we can add in a future PR.

**To do:**
- [x] Check file upload feature
- [x] Test for other functionality that may error
- [x] Add unit test for decorator

### CODE REVIEW

File: hnn_core/gui/_viz_manager.py
Comment (ntolley) [inline hnn_core/gui/_viz_manager.py:742]: Need to think of one line comment on ipywidgets figure linking logic

**Comment by gtdang** [comment]
@ntolley Seems like the macOS tests are passing now. It's failing on docs and link check? Are you familiar with these errors?

**Comment by ntolley** [comment]
New error for me as well @gtdang 

Looks like it's a weird consequence of one of the version changes with the `ipywidgets`?

**Comment by ntolley** [comment]
@gtdang I did some digging and I think the answer is to add `ipykernel` as an explicit dependency:

https://stackoverflow.com/questions/76855161/no-such-kernel-named-python3-ipywidgets-ipykernel-dependency-issue

https://github.com/jupyter-widgets/ipywidgets/issues/3731

**Comment by gtdang** [comment]
@ntolley adding ipykernel as a GUI dependency seemed to do the trick! 

Now the docs build is timing out as it's taking longer than 10mins. Does this tend to happen? If so, we could try [increasing the timeout limit ](https://support.circleci.com/hc/en-us/articles/360007188574-Build-has-Hit-Timeout-Limit#:~:text=CircleCI%20has%20a%20built%2Din,will%20be%20canceled%20and%20stopped.)from the default 10m.

**Comment by gtdang** [comment]
> @ntolley adding ipykernel as a GUI dependency seemed to do the trick!
> 
> Now the docs build is timing out as it's taking longer than 10mins. Does this tend to happen? If so, we could try [increasing the timeout limit ](https://support.circleci.com/hc/en-us/articles/360007188574-Build-has-Hit-Timeout-Limit#:~:text=CircleCI%20has%20a%20built%2Din,will%20be%20canceled%20and%20stopped.)from the default 10m.

Hm never mind. It seems like the GUI captures are not rendering for creating the docs. I'll have to dig into why this is happening with the updated dependencies.

**Comment by gtdang** [comment]
@ntolley I updated the circleci config file to bump the build timeout to 20mins. It passed this time however looking the logs the build took <2min... So I'm thinking that change wasn't necessary and the timeout was just some inconsistency/stability issue on the vm. But I guess it doesn't hurt to keep it at 20mins? I'm not sure what plan you are on and if runtime is limited.

File: .circleci/config.yml
Comment (ntolley) [inline .circleci/config.yml:65]: I think it'd be best to keep this at 10 min if it's not the actual issue. In general we're trying to avoid having our documentation take a long time to build until we have a clear plan in place for something like the GUI tutorials where long build times are a bit unavoidable.

File: setup.py
Comment (ntolley) [inline setup.py:109]: is it not necessary to explicitly specify `voila`?

File: hnn_core/tests/test_gui.py
Comment (ntolley) [inline hnn_core/tests/test_gui.py:453]: Will this work in a `pytest` context manager? Something like

https://github.com/jonescompneurolab/hnn-core/blob/34d972ce8a57d4a04a281a312884f7daa597ec49/hnn_core/tests/test_gui.py#L103C18-L103C18

Comment (gtdang) [reply hnn_core/tests/test_gui.py:453]: Ooo I didn't know about that. That's much cleaner thanks!

File: setup.py
Comment (gtdang) [reply setup.py:109]: Hmm... that's a good point. I'll look into this. I'm curious how the test builds are getting voila if it's not specified...

File: .circleci/config.yml
Comment (gtdang) [reply .circleci/config.yml:65]: Ok sounds good

File: setup.py
Comment (gtdang) [reply setup.py:109]: So there's currently no test for launching the GUI stand-alone. We're only testing the widget components, which is all independent of voila. As for ipympl, we also do not have a test to make sure the plots have the interaction options. 

At the bare minimum we could just do an import test. This would only ensure that we don't make a mistake in the setup file (like above 😆).

However a more thorough test would test if the stand-alone GUI actually launches and renders correctly. This is a bit more complicated and would involve implementing voila tests using additional dependencies like [solara](https://solara.dev/docs/howto/testing) and [playwright](https://playwright.dev/python/docs/api/class-playwright).

Comment (gtdang) [reply setup.py:109]: It would also be a good idea to create some specified dependency groups for docs, test, and dev. This would allow contributors to install all the development dependencies listed in the contributions guide with `pip install -e '.[dev]'` The setup file would have a dict construction as follows: 

```python
extras = {
              'gui': ['ipywidgets>=8.0.0', 'ipykernel', 'voila', 'ipympl',
                      'joblib'
                      ],
              'opt': ['scikit-learn'],
              'test': ['flake8', 'pytest', 'pytest-cov',
                       'solara', 'playwright'
                       ],
              'docs': ['psutil', 'mne', 'sphinx', 'nbsphinx', 'sphinx-gallery',
                       'sphinx_bootstrap_theme', 'sphinx-copybutton', 'pillow',
                       'numpydoc',
                       ]
          }
extras['dev'] = extras['test'] + extras['docs']
```

File: hnn_core/gui/_viz_manager.py
Comment (jasmainak) [inline hnn_core/gui/_viz_manager.py:104]: Check PEP 257 convention ... need a one-line description first ... longer description may follow later.

File: hnn_core/gui/_viz_manager.py
Comment (jasmainak) [inline hnn_core/gui/_viz_manager.py:102]: we don't do type annotations ... I think it will have to be a repo-wide change, let's stick to repo conventions for now to keep things consistent

File: hnn_core/gui/_viz_manager.py
Comment (jasmainak) [reply hnn_core/gui/_viz_manager.py:104]: "Wrapper function" as in decorator, right?

File: hnn_core/gui/_viz_manager.py
Comment (jasmainak) [inline hnn_core/gui/_viz_manager.py:603]: backslashes are not preferred ... would suggest using paranthesis if possible

File: hnn_core/gui/_viz_manager.py
Comment (jasmainak) [reply hnn_core/gui/_viz_manager.py:749]: sorry I missed the discussion on this, is it a bugfix or something related to the new version of ipywidgets?

File: hnn_core/tests/test_gui.py
Comment (jasmainak) [inline hnn_core/tests/test_gui.py:10]: paranthesis instead of backslash

File: hnn_core/tests/test_gui.py
Comment (jasmainak) [inline hnn_core/tests/test_gui.py:425]: why not use `VizManager` class directly for testing? what's the bottleneck?

**Comment by jasmainak** [comment]
> But I guess it doesn't hurt to keep it at 20mins?

we need to find a solution to this ... @gtdang if you are willing to help that would be great. IMO, the GUI docs should be run separately. Currently running `make html-noplot` also runs the GUI examples ... it's a bug

we need a dedicated machine for running the GUI docs because they are slow ... either on a paid plan on CircleCI or through the Brown infrastructure. Not sure how we'd link it to pull requests if we do it on the Brown infrastructure though.

File: setup.py
Comment (jasmainak) [reply setup.py]: > However a more thorough test would test if the stand-alone GUI actually launches and renders correctly. This is a bit more complicated and would involve implementing voila tests using additional dependencies like [solara](https://solara.dev/docs/howto/testing) and [playwright](https://playwright.dev/python/docs/api/class-playwright).

This is interesting! Can you dig a bit and research what makes more sense? From a cursory look, Solara looked appropriate although I'm not sure how engaged the community is. Is it easy to install etc. We want something that's lightweight and easy for new developers. We often get early Masters students who do not have a lot of knowledge of software testing but want to contribute to HNN. The more we document and make the process of becoming a new developer easier, the better.

Comment (jasmainak) [reply setup.py]: > It would also be a good idea to create some specified dependency groups for docs, test, and dev. 

agreed, as long we document everything properly! We should update our contributing guide as you go through the process:

https://jonescompneurolab.github.io/hnn-core/stable/contributing.html

**Comment by jasmainak** [comment]
> But I guess it doesn't hurt to keep it at 20mins?

Let's merge this repo with the 10 min limit and try to find a long-term solution in separate pull requests

File: hnn_core/tests/test_gui.py
Comment (gtdang) [reply hnn_core/tests/test_gui.py:425]: It's mostly out of principle to make the test small and clear so that the scope of the test is just the functionality of the decorator, which makes it easier to detect if the a bug is coming from within the function vs externally. 

The full functionality tests of the VizManager like`test_gui_add_figure` would already catch/debug bugs that occur in VizManager and its use of the decorator.

**Comment by gtdang** [comment]
> > But I guess it doesn't hurt to keep it at 20mins?
> 
> Let's merge this PR with the 10 min limit and try to find a long-term solution in separate pull requests

We actually found another issue on Wednesday with the `.capture()` method no longer rendering the GUI--an issue for the GUI docs/tutorials. After some digging it's related to https://github.com/jupyter-widgets/ipywidgets/issues/3871. So we'll need to find an alternative solution for capturing the state of the GUI and displaying it. Should we hold off on merging until that is figured out?

**Comment by jasmainak** [comment]
ughgh ... any workarounds for now?

maybe: https://html2canvas.hertzen.com/ ?

**Comment by ntolley** [comment]
Not a bad option @jasmainak !

Truthfully I think it may be best to keep this PR focused on making the GUI functional, and in a separate one either fix the docs or test a new solution all together

**Comment by gtdang** [comment]
> Not a bad option @jasmainak !
> 
> Truthfully I think it may be best to keep this PR focused on making the GUI functional, and in a separate one either fix the docs or test a new solution all together

I agree. I think the docs solution will be in a separate PR. I'm not in love the the current implementation of the GUI tutorials (even when it is working properly). It's a bit confusing and resource intensive for for the end user. I think screenshots would more clear for the user, although it has the tradeoff being less a streamlined process for writers. We can discuss more when we meet this week.

**Comment by jasmainak** [comment]
will `pyautogui` require `xvfb` or a virtual screen to make it work in the CIs? My concern is that exact coordinates for the CI may be difficult to guess ... but I'm open to anything that works and is a stable solution

**Comment by gtdang** [comment]
> will `pyautogui` require `xvfb` or a virtual screen to make it work in the CIs? My concern is that exact coordinates for the CI may be difficult to guess ... but I'm open to anything that works and is a stable solution

Yes it would be a 2 step process of first running the simulation and grabbing the screenshots and then creating the documentation using the images. It shouldn't need to be done every time we regenerate docs though, only if the UI changes. You can actually give it images of buttons/tabs for it to find the coordinates to click. The images should also have the mouse cursor to highlight context. 

For html2canvas... I would be against adding a js runtime dependency just for screen caps. Also the execution would be pretty difficult because you would have to invoke the js after each cell execution... not at the end of the full notebook execution.

**Comment by ntolley** [comment]
@jasmainak I'm good with going ahead and merging this PR as I think in the long term we are going to move to a screenshot method for tutorials (either automated or manual). In any case I think the purpose of this PR is achieved, the next one to work on is #651

Merge button is yours if you're good with its current state!

**Review by ntolley — APPROVED** [review]

File: setup.py
Comment (jasmainak) [inline setup.py:109]: I think you'll need to update README:

https://jonescompneurolab.github.io/hnn-core/stable/index.html

let me know once you fix it and I'll go ahead and merge @gtdang !

Comment (gtdang) [reply setup.py:109]: Thanks! Updated!

**Comment by jasmainak** [comment]
Thanks @gtdang ! Congratulations on your first merged PR 🥳 🍺 🎉

**Pull request merged**
